src/backend/distributed/utils/replication_origin_session_utils.c

```diff
@@ -0,0 +1,222 @@
+/*-------------------------------------------------------------------------
+ *
+ * replication_origin_session_utils.c
+ *   Functions for managing replication origin session.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "distributed/replication_origin_session_utils.h"
+#include "distributed/remote_commands.h"
+#include "distributed/metadata_cache.h"
+
+static bool isReplicationOriginSessionSetup(MultiConnection *connection);
+static bool isReplicationOriginCreated(MultiConnection *connection, char *originName,
+									   RepOriginId *originId);
+static RepOriginId ReplicationOriginSessionCreate(MultiConnection *connection,
+												  char *originName);
+static void ReplicationOriginSessionSetupHelper(MultiConnection *connection,
+												RepOriginId originId, char *originName);
+static bool ExecuteRemoteCommandAndCheckResult(MultiConnection *connection, char *command,
+											   char *expected);
+
+/* ReplicationOriginSessionSetup sets up a new replication origin session in a
+ * local or remote session depending on the useLocalCopy flag. If useLocalCopy
+ * is set, a local replication origin session is setup, otherwise a remote
+ * replication origin session is setup to the destination node.
+ */
+void
+ReplicationOriginSessionSetup(MultiConnection *connection)
+{
+	if (!isReplicationOriginSessionSetup(connection))
+	{
+		int localid = GetLocalNodeId();
+		RepOriginId originId = InvalidRepOriginId;
+		StringInfo originNameString = makeStringInfo();
+		appendStringInfo(originNameString, "citus_internal_%d", localid);
+		if (!isReplicationOriginCreated(connection, originNameString->data, &originId))
+		{
+			originId = ReplicationOriginSessionCreate(connection, originNameString->data);
+		}
+		ReplicationOriginSessionSetupHelper(connection, originId, originNameString->data);
+	}
+}
+
+
+/* ReplicationOriginSessionReset resets the replication origin session in a
+ * local or remote session depending on the useLocalCopy flag.
+ */
+void
+ReplicationOriginSessionReset(MultiConnection *connection)
+{
+	if (connection == NULL)
+	{
+		/*Reset Replication Origin in local session */
+		if (replorigin_session_origin != InvalidRepOriginId)
+		{
+			replorigin_session_reset();
+			replorigin_session_origin = InvalidRepOriginId;
```

Comment:
for both of these cases, but especially the local one, we'll need to make sure that the replication origin gets reset even in case of an error (e.g. out of memory), such that we do not skip all the subsequent writes in the session.

Reply:
Added this "ResetReplicationOriginSession()"  calls to the abourt cases in  transaction_management.c